diff -ruN rustc-1.30.1-src/src/vendor/openssl-sys/build/main.rs arustc/src/vendor/openssl-sys/build/main.rs
--- rustc-1.30.1-src/src/vendor/openssl-sys/build/main.rs	2018-11-07 08:07:14.000000000 +0300
+++ arustc/src/vendor/openssl-sys/build/main.rs	2018-11-13 17:44:13.925926760 +0300
@@ -1,9 +1,9 @@
 extern crate cc;
+#[cfg(feature = "vendored")]
+extern crate openssl_src;
 extern crate pkg_config;
 #[cfg(target_env = "msvc")]
 extern crate vcpkg;
-#[cfg(feature = "vendored")]
-extern crate openssl_src;
 
 use std::collections::HashSet;
 use std::env;
@@ -29,6 +29,7 @@
     "OPENSSL_NO_SRP",
     "OPENSSL_NO_SSL3_METHOD",
     "OPENSSL_NO_TLSEXT",
+    "OPENSSL_NO_STDIO",
 ];
 
 enum Version {
@@ -102,20 +103,23 @@
 
 #[cfg(feature = "vendored")]
 mod imp {
-	use std::path::PathBuf;
-	use openssl_src;
+    use openssl_src;
+    use std::path::PathBuf;
 
-	pub fn get_openssl(_target: &str) -> (PathBuf, PathBuf) {
-		let artifacts = openssl_src::Build::new().build();
-		(artifacts.lib_dir().to_path_buf(), artifacts.include_dir().to_path_buf())
-	}
+    pub fn get_openssl(_target: &str) -> (PathBuf, PathBuf) {
+        let artifacts = openssl_src::Build::new().build();
+        (
+            artifacts.lib_dir().to_path_buf(),
+            artifacts.include_dir().to_path_buf(),
+        )
+    }
 }
 
 #[cfg(not(feature = "vendored"))]
 mod imp {
     use pkg_config;
-    use std::path::{Path, PathBuf};
     use std::ffi::OsString;
+    use std::path::{Path, PathBuf};
     use std::process::{self, Command};
 
     use super::env;
@@ -177,23 +181,18 @@
 
         let mut msg = format!(
             "
-
 Could not find directory of OpenSSL installation, and this `-sys` crate cannot
 proceed without this knowledge. If OpenSSL is installed and this crate had
 trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
 compilation process.
-
 Make sure you also have the development packages of openssl installed.
 For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.
-
 If you're in a situation where you think the directory *should* be found
 automatically, please open a bug at https://github.com/sfackler/rust-openssl
 and include information about your system as well as this message.
-
     $HOST = {}
     $TARGET = {}
     openssl-sys = {}
-
 ",
             host,
             target,
@@ -205,16 +204,12 @@
             if system.exists() {
                 msg.push_str(&format!(
                     "
-
 It looks like you're compiling on macOS, where the system contains a version of
 OpenSSL 0.9.8. This crate no longer supports OpenSSL 0.9.8.
-
 As a consumer of this crate, you can fix this error by using Homebrew to
 install the `openssl` package, or as a maintainer you can use the openssl-sys
 0.7 crate for support with OpenSSL 0.9.8.
-
 Unfortunately though the compile cannot continue, so aborting.
-
 "
                 ));
             }
@@ -228,7 +223,6 @@
 requires the `pkg-config` utility to find OpenSSL but unfortunately `pkg-config`
 could not be found. If you have OpenSSL installed you can likely fix this by
 installing `pkg-config`.
-
 "
                 ));
             }
@@ -239,11 +233,8 @@
                 "
 It looks like you're compiling for MinGW but you may not have either OpenSSL or
 pkg-config installed. You can install these two dependencies with:
-
     pacman -S openssl-devel pkg-config
-
 and try building this crate again.
-
 "
             ));
         }
@@ -255,9 +246,7 @@
 installation. If there isn't one installed then you can try the rust-openssl
 README for more information about how to download precompiled binaries of
 OpenSSL:
-
     https://github.com/sfackler/rust-openssl#windows
-
 "
             ));
         }
@@ -390,12 +379,9 @@
         "\
 #include <openssl/opensslv.h>
 #include <openssl/opensslconf.h>
-
 #define VERSION2(n, v) RUST_VERSION_ ## n ## _ ## v
 #define VERSION(n, v) VERSION2(n, v)
-
 VERSION(OPENSSL, OPENSSL_VERSION_NUMBER)
-
 #ifdef LIBRESSL_VERSION_NUMBER
 VERSION(LIBRESSL, LIBRESSL_VERSION_NUMBER)
 #endif
@@ -429,22 +415,17 @@
                 "
 Header expansion error:
 {:?}
-
 Failed to find OpenSSL development headers.
-
 You can try fixing this setting the `OPENSSL_DIR` environment variable
 pointing to your OpenSSL installation or installing OpenSSL headers package
 specific to your distribution:
-
     # On Ubuntu
     sudo apt-get install libssl-dev
     # On Arch Linux
     sudo pacman -S openssl
     # On Fedora
     sudo dnf install openssl-devel
-
 See rust-openssl README for more information:
-
     https://github.com/sfackler/rust-openssl#linux
 ",
                 e
@@ -497,6 +478,9 @@
             (6, 2) => ('6', '2'),
             (6, _) => ('6', 'x'),
             (7, _) => ('7', 'x'),
+            (8, 0) => ('8', '0'),
+            (8, 1) => ('8', '1'),
+            (8, _) => ('8', 'x'),
             _ => version_error(),
         };
 
@@ -535,11 +519,9 @@
 fn version_error() -> ! {
     panic!(
         "
-
 This crate is only compatible with OpenSSL 1.0.1 through 1.1.1, or LibreSSL 2.5
-through 2.7, but a different version of OpenSSL was found. The build is now aborting
+through 2.8.1, but a different version of OpenSSL was found. The build is now aborting
 due to this version mismatch.
-
 "
     );
 }
